File: guides/micronaut-streamed-file-and-reactor-streaming-http-client/java/src/main/java/example/micronaut/HomeController.java
package example.micronaut;

import io.micronaut.context.exceptions.ConfigurationException;
import io.micronaut.core.io.buffer.ByteBuffer;
import io.micronaut.core.io.buffer.ReferenceCounted;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.reactor.http.client.ReactorStreamingHttpClient;
import jakarta.annotation.PreDestroy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Flux;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;

@Controller // <1>
class HomeController implements AutoCloseable {
    private static final Logger LOG = LoggerFactory.getLogger(HomeController.class);
    private static final URI DEFAULT_URI = URI.create("https://guides.micronaut.io/micronaut5K.png");

    private final ReactorStreamingHttpClient reactorStreamingHttpClient;

    HomeController() {
        String urlStr = "https://guides.micronaut.io/";
        URL url;
        try {
            url = new URL(urlStr);
        } catch (MalformedURLException e) {
            throw new ConfigurationException("malformed URL" + urlStr);
        }
        this.reactorStreamingHttpClient = ReactorStreamingHttpClient.create(url); // <2>
    }

    @Get // <3>
    Flux<ByteBuffer<?>> download() {
        HttpRequest<?> request = HttpRequest.GET(DEFAULT_URI);
        return reactorStreamingHttpClient.dataStream(request).doOnNext(bb -> {
            if (bb instanceof ReferenceCounted rc) {
                rc.retain();
            }
        }); // <4>
    }

    @PreDestroy // <5>
    @Override
    public void close() {
        if (reactorStreamingHttpClient != null) {
            reactorStreamingHttpClient.close();
        }
    }
}

File: guides/micronaut-reactor-streaming-http-client/python/src/example/micronaut/home_controller.py
from java.net import URI, URL
from java.time import Duration
from jakarta.annotation import PreDestroy
from micronaut.core.io.buffer import ReferenceCounted
from micronaut.http import HttpRequest, MediaType
from micronaut.http.annotation import Controller, Get, Produces
from micronaut.http.client import DefaultHttpClientConfiguration
from micronaut.reactor.http.client import ReactorStreamingHttpClient
from org.reactivestreams import Publisher
from reactor.core.publisher import Flux

DEFAULT_URI = URI.create(
    "https://raw.githubusercontent.com/micronaut-projects/micronaut-guides/master/guides/micronaut-reactor-streaming-http-client/src/test/resources/micronaut5K.png"
)


@Controller  # <1>
class HomeController:
    def __init__(self):
        configuration = DefaultHttpClientConfiguration()
        configuration.setReadTimeout(Duration.ofSeconds(30))
        self.reactor_streaming_http_client = ReactorStreamingHttpClient.create(
            URL("https://guides.micronaut.io/"), configuration
        )  # <2>

    @Produces(MediaType.IMAGE_PNG)
    @Get  # <3>
    def download(self) -> Publisher:
        request = HttpRequest.GET(DEFAULT_URI)
        return Flux.from_(self.reactor_streaming_http_client.dataStream(request)).doOnNext(
            retain_reference_counted
        )  # <4>

    @PreDestroy  # <5>
    def close(self) -> None:
        if self.reactor_streaming_http_client is not None:
            self.reactor_streaming_http_client.close()


def retain_reference_counted(byte_buffer) -> None:
    if isinstance(byte_buffer, ReferenceCounted):
        byte_buffer.retain()
